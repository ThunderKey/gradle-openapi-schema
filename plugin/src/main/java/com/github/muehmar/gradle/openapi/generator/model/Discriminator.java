package com.github.muehmar.gradle.openapi.generator.model;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class Discriminator {
  private final Name propertyName;
  private final Optional<Map<String, PojoName>> mapping;

  private Discriminator(Name propertyName, Optional<Map<String, PojoName>> mapping) {
    this.propertyName = propertyName;
    this.mapping = mapping;
  }

  public static Discriminator fromPropertyName(Name propertyName) {
    return new Discriminator(propertyName, Optional.empty());
  }

  public static Discriminator fromPropertyNameAndMapping(
      Name propertyName, Map<String, PojoName> mapping) {
    return new Discriminator(propertyName, Optional.of(mapping));
  }

  public Discriminator withMapping(Optional<Map<String, PojoName>> mapping) {
    return new Discriminator(propertyName, mapping);
  }

  public Name getPropertyName() {
    return propertyName;
  }

  public String getValueForPojoName(PojoName pojoName) {
    return mapping.orElse(Collections.emptyMap()).entrySet().stream()
        .filter(e -> e.getValue().equals(pojoName))
        .findFirst()
        .map(Map.Entry::getKey)
        .orElse(pojoName.getName().asString());
  }
}
