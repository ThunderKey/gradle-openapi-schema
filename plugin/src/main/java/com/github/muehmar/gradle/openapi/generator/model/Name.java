package com.github.muehmar.gradle.openapi.generator.model;

import java.util.function.UnaryOperator;
import lombok.EqualsAndHashCode;

/** Name of a pojo or a pojo member. */
@EqualsAndHashCode
public class Name {
  private final String value;

  private Name(String val) {
    this.value = val;
  }

  public static Name ofString(String val) {
    if (val == null || val.trim().isEmpty()) {
      throw new IllegalArgumentException("A name must not be null or empty");
    }

    return new Name(val);
  }

  public String asString() {
    return value;
  }

  public Name map(UnaryOperator<String> f) {
    return Name.ofString(f.apply(value));
  }

  public Name append(String append) {
    return new Name(value + append);
  }

  public Name append(Name append) {
    return new Name(value + append.value);
  }

  public Name prefix(String prefix) {
    return new Name(prefix + value);
  }

  public SuffixedName suffix(String suffix) {
    return SuffixedName.ofNameAndSuffix(this, suffix);
  }

  public boolean equalsIgnoreCase(Name other) {
    return value.equalsIgnoreCase(other.value);
  }

  public Name startUpperCase() {
    return map(s -> s.substring(0, 1).toUpperCase() + s.substring(1));
  }

  @Override
  public String toString() {
    return asString();
  }
}
