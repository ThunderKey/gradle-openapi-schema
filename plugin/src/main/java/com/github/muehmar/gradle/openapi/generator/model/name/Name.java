package com.github.muehmar.gradle.openapi.generator.model.name;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import java.util.function.UnaryOperator;
import lombok.EqualsAndHashCode;

/** General-purpose name. */
@EqualsAndHashCode
public class Name {
  private final String value;

  private Name(String val) {
    this.value = val;
  }

  public static Name ofString(String val) {
    if (val == null || val.trim().isEmpty()) {
      throw new IllegalArgumentException("A name must not be null or empty");
    }

    return new Name(val);
  }

  public String asString() {
    return value;
  }

  public Name map(UnaryOperator<String> f) {
    return Name.ofString(f.apply(value));
  }

  public Name append(String append) {
    return new Name(value + append);
  }

  public Name append(Name append) {
    return new Name(value + append.value);
  }

  public Name prefix(String prefix) {
    return new Name(prefix + value);
  }

  public boolean equalsIgnoreCase(Name other) {
    return value.equalsIgnoreCase(other.value);
  }

  public Name startUpperCase() {
    return map(s -> s.substring(0, 1).toUpperCase() + s.substring(1));
  }

  public Name startLowerCase() {
    return map(s -> s.substring(0, 1).toLowerCase() + s.substring(1));
  }

  public boolean contains(CharSequence s) {
    return value.contains(s);
  }

  public boolean startsNotWith(String prefix) {
    return not(value.startsWith(prefix));
  }

  @Override
  public String toString() {
    return asString();
  }
}
