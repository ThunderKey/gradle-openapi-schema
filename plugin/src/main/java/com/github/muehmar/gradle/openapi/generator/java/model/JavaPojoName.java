package com.github.muehmar.gradle.openapi.generator.java.model;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.generator.model.name.PojoName;
import lombok.EqualsAndHashCode;

@EqualsAndHashCode
public class JavaPojoName {
  private final String name;
  private final String suffix;

  private JavaPojoName(String name, String suffix) {
    this.name = name;
    this.suffix = suffix;
  }

  public static JavaPojoName wrap(PojoName pojoName) {
    return new JavaPojoName(pojoName.getName().asString(), pojoName.getSuffix());
  }

  public static JavaPojoName fromNameAndSuffix(String name, String suffix) {
    return new JavaPojoName(name, suffix);
  }

  public PojoName asPojoName() {
    return PojoName.ofNameAndSuffix(name, suffix);
  }

  public JavaPojoName appendToName(String append) {
    return new JavaPojoName(name.concat(append), suffix);
  }

  public JavaName asJavaName() {
    return JavaName.fromName(Name.ofString(asString()));
  }

  public JavaIdentifier asIdentifier() {
    return asJavaName().asIdentifier();
  }

  public String asString() {
    return name.concat(suffix);
  }

  @Override
  public String toString() {
    return asString();
  }
}
