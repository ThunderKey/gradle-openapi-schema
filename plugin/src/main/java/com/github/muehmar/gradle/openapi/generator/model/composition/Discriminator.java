package com.github.muehmar.gradle.openapi.generator.model.composition;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class Discriminator {
  private final Name propertyName;
  private final DiscriminatorType type;
  private final Optional<Map<String, Name>> mapping;

  private Discriminator(
      Name propertyName, DiscriminatorType type, Optional<Map<String, Name>> mapping) {
    this.propertyName = propertyName;
    this.type = type;
    this.mapping = mapping;
  }

  public static Discriminator typeDiscriminator(
      UntypedDiscriminator untypedDiscriminator, DiscriminatorType type) {
    return new Discriminator(
        untypedDiscriminator.getPropertyName(), type, untypedDiscriminator.getMapping());
  }

  public Name getPropertyName() {
    return propertyName;
  }

  public DiscriminatorType getType() {
    return type;
  }

  public String getValueForSchemaName(Name schemaName) {
    return mapping.orElse(Collections.emptyMap()).entrySet().stream()
        .filter(e -> e.getValue().equals(schemaName))
        .findFirst()
        .map(Map.Entry::getKey)
        .orElse(schemaName.asString());
  }
}
