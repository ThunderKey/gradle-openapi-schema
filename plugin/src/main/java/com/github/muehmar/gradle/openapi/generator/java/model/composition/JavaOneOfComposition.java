package com.github.muehmar.gradle.openapi.generator.java.model.composition;

import static com.github.muehmar.gradle.openapi.generator.java.model.composition.Assertion.assertAllObjectPojos;

import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.composition.OneOfComposition;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaOneOfComposition {
  private final NonEmptyList<JavaObjectPojo> pojos;
  private final Optional<Discriminator> discriminator;

  private JavaOneOfComposition(
      NonEmptyList<JavaPojo> pojos, Optional<Discriminator> discriminator) {
    this.pojos = assertAllObjectPojos(pojos);
    this.discriminator = discriminator;
  }

  public static JavaOneOfComposition wrap(
      OneOfComposition oneOfComposition, PojoType type, TypeMappings typeMappings) {
    return new JavaOneOfComposition(
        oneOfComposition
            .getPojos()
            .map(pojo -> JavaPojo.wrap(pojo, typeMappings))
            .map(result -> result.getTypeOrDefault(type)),
        oneOfComposition.getDiscriminator());
  }

  public static JavaOneOfComposition fromPojos(NonEmptyList<JavaPojo> pojos) {
    return new JavaOneOfComposition(pojos, Optional.empty());
  }

  public static JavaOneOfComposition fromPojosAndDiscriminator(
      NonEmptyList<JavaPojo> pojos, Discriminator discriminator) {
    return new JavaOneOfComposition(pojos, Optional.of(discriminator));
  }

  public NonEmptyList<JavaObjectPojo> getPojos() {
    return pojos;
  }

  public PList<JavaPojoMember> getMembers() {
    return pojos
        .toPList()
        .flatMap(JavaObjectPojo::getAllMembersForComposition)
        .map(JavaPojoMember::asOneOfMember);
  }

  public Optional<Discriminator> getDiscriminator() {
    return discriminator;
  }
}
