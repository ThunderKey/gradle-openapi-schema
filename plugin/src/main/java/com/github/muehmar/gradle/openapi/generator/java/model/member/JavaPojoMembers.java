package com.github.muehmar.gradle.openapi.generator.java.model.member;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import java.util.function.UnaryOperator;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * The members of a {@link JavaObjectPojo}. This container ensures, that the same property does not
 * exist twice. Adding a property which is already present will not add a new property/member, it
 * will keep the current property but maybe with lower restrictions concerning {@link Nullability}
 * and {@link Necessity}.
 */
@EqualsAndHashCode
@ToString
public class JavaPojoMembers {
  private final PList<JavaPojoMember> members;

  private JavaPojoMembers(PList<JavaPojoMember> members) {
    this.members = members;
  }

  public static JavaPojoMembers empty() {
    return new JavaPojoMembers(PList.empty());
  }

  public static JavaPojoMembers fromMembers(PList<JavaPojoMember> members) {
    return JavaPojoMembers.empty().add(members);
  }

  public JavaPojoMembers add(JavaPojoMember memberToAdd) {
    final boolean hasMemberWithSameKey =
        members.exists(member -> member.getMemberKey().equals(memberToAdd.getMemberKey()));
    if (hasMemberWithSameKey) {
      final PList<JavaPojoMember> newMembers =
          members.map(member -> member.mergeToLeastRestrictive(memberToAdd).orElse(member));
      return new JavaPojoMembers(newMembers);
    } else {
      return new JavaPojoMembers(members.add(memberToAdd));
    }
  }

  public JavaPojoMembers add(PList<JavaPojoMember> membersToAdd) {
    return membersToAdd.foldLeft(this, JavaPojoMembers::add);
  }

  public JavaPojoMembers add(JavaPojoMembers membersToAdd) {
    return add(membersToAdd.asList());
  }

  public JavaPojoMembers map(UnaryOperator<JavaPojoMember> f) {
    return new JavaPojoMembers(members.map(f));
  }

  public PList<JavaPojoMember> asList() {
    return members;
  }

  public boolean isEmpty() {
    return members.isEmpty();
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public int getRequiredMemberCount() {
    return members.filter(JavaPojoMember::isRequired).size();
  }
}
