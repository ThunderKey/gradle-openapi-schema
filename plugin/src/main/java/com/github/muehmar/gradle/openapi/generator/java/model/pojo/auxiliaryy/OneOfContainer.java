package com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy;

import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.anyTypeAllowed;

import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import lombok.Value;

@Value
public class OneOfContainer {
  private static final String NAME_SUFFIX = "OneOfContainer";
  JavaPojoName pojoName;
  JavaOneOfComposition composition;

  public JavaPojoName getContainerName() {
    return pojoName.appendToName(NAME_SUFFIX);
  }

  public MemberGenerator.MemberContent memberContent() {
    return MemberContentBuilder.fullMemberContentBuilder()
        .isArrayPojo(false)
        .members(composition.getMembers())
        .additionalProperties(anyTypeAllowed())
        .build();
  }
}
