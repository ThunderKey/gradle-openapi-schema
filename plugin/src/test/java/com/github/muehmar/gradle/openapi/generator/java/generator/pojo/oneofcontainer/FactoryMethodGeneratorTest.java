package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.oneofcontainer;

import static com.github.muehmar.gradle.openapi.SnapshotUtil.writerSnapshot;
import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.oneofcontainer.FactoryMethodGenerator.factoryMethodGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo1;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo2;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import ch.bluecare.commons.data.NonEmptyList;
import com.github.muehmar.gradle.openapi.IntellijDiffSnapshotTestExtension;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy.OneOfContainer;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.HashMap;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith({SnapshotExtension.class, IntellijDiffSnapshotTestExtension.class})
class FactoryMethodGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("allNecessityAndNullabilityVariantsPojo")
  void generate_when_allNecessityAndNullabilityVariantsPojo_then_correctOutput() {
    final Generator<OneOfContainer, PojoSettings> generator = factoryMethodGenerator();

    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfComposition.fromPojos(
            NonEmptyList.of(sampleObjectPojo1(), JavaPojos.allNecessityAndNullabilityVariants()));
    final OneOfContainer oneOfContainer =
        new OneOfContainer(JavaPojoName.fromNameAndSuffix("Object", "Dto"), javaOneOfComposition);
    final Writer writer =
        generator.generate(
            oneOfContainer, TestPojoSettings.defaultSettings(), Writer.createDefault());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("discriminatorWithoutMapping")
  void generate_when_discriminatorWithoutMapping_then_correctOutput() {
    final Generator<OneOfContainer, PojoSettings> generator = factoryMethodGenerator();

    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfComposition.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()),
            Discriminator.fromPropertyName(JavaPojoMembers.requiredString().getName().asName()));
    final OneOfContainer oneOfContainer =
        new OneOfContainer(JavaPojoName.fromNameAndSuffix("Object", "Dto"), javaOneOfComposition);
    final Writer writer =
        generator.generate(
            oneOfContainer, TestPojoSettings.defaultSettings(), Writer.createDefault());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("discriminatorWithMapping")
  void generate_when_discriminatorWithMapping_then_correctOutput() {
    final Generator<OneOfContainer, PojoSettings> generator = factoryMethodGenerator();
    final HashMap<String, Name> mapping = new HashMap<>();
    mapping.put("Custom1", sampleObjectPojo1().getSchemaName().asName());
    mapping.put("Custom2", sampleObjectPojo2().getSchemaName().asName());

    final Discriminator discriminator =
        Discriminator.fromPropertyNameAndMapping(
            JavaPojoMembers.requiredString().getName().asName(), mapping);
    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfComposition.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()), discriminator);
    final OneOfContainer oneOfContainer =
        new OneOfContainer(JavaPojoName.fromNameAndSuffix("Object", "Dto"), javaOneOfComposition);
    final Writer writer =
        generator.generate(
            oneOfContainer, TestPojoSettings.defaultSettings(), Writer.createDefault());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
