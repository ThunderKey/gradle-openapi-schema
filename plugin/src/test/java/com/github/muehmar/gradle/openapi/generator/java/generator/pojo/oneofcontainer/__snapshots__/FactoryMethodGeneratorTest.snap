allNecessityAndNullabilityVariantsPojo=[
.
.
/**
 * Creates an instance of {@link ObjectOneOfContainerDto} from a {@link
 * SampleObjectPojo1Dto}.
 */
public static ObjectOneOfContainerDto fromSampleObjectPojo1(SampleObjectPojo1Dto dto) {
  return new ObjectOneOfContainerDto(
    dto.getStringVal(),
    dto.getIntVal(),
    dto.getDoubleVal(),
    null,
    null,
    false,
    null,
    null,
    false,
    dto.getAdditionalProperties()
  );
}

/**
 * Creates an instance of {@link ObjectOneOfContainerDto} from a {@link
 * NecessityAndNullabilityDto}.
 */
public static ObjectOneOfContainerDto fromNecessityAndNullability(NecessityAndNullabilityDto dto) {
  return new ObjectOneOfContainerDto(
    null,
    null,
    null,
    dto.getRequiredStringVal(),
    dto.getRequiredNullableStringValOr(null),
    true,
    dto.getOptionalStringValOr(null),
    dto.getOptionalNullableStringVal().onValue(val -> val).onNull(() -> null).onAbsent(() -> null),
    dto.getOptionalNullableStringVal().onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false),
    dto.getAdditionalProperties()
  );
}
]


discriminatorWithMapping=[
.
.
/**
 * Creates an instance of {@link ObjectOneOfContainerDto} from a {@link
 * SampleObjectPojo1Dto}.
 */
public static ObjectOneOfContainerDto fromSampleObjectPojo1(SampleObjectPojo1Dto dto) {
  return new ObjectOneOfContainerDto(
    "Custom1",
    dto.getIntVal(),
    dto.getDoubleVal(),
    null,
    null,
    dto.getAdditionalProperties()
  );
}

/**
 * Creates an instance of {@link ObjectOneOfContainerDto} from a {@link
 * SampleObjectPojo2Dto}.
 */
public static ObjectOneOfContainerDto fromSampleObjectPojo2(SampleObjectPojo2Dto dto) {
  return new ObjectOneOfContainerDto(
    "Custom2",
    null,
    null,
    dto.getBirthdate(),
    dto.getEmail(),
    dto.getAdditionalProperties()
  );
}
]


discriminatorWithoutMapping=[
.
.
/**
 * Creates an instance of {@link ObjectOneOfContainerDto} from a {@link
 * SampleObjectPojo1Dto}.
 */
public static ObjectOneOfContainerDto fromSampleObjectPojo1(SampleObjectPojo1Dto dto) {
  return new ObjectOneOfContainerDto(
    "SampleObjectPojo1",
    dto.getIntVal(),
    dto.getDoubleVal(),
    null,
    null,
    dto.getAdditionalProperties()
  );
}

/**
 * Creates an instance of {@link ObjectOneOfContainerDto} from a {@link
 * SampleObjectPojo2Dto}.
 */
public static ObjectOneOfContainerDto fromSampleObjectPojo2(SampleObjectPojo2Dto dto) {
  return new ObjectOneOfContainerDto(
    "SampleObjectPojo2",
    null,
    null,
    dto.getBirthdate(),
    dto.getEmail(),
    dto.getAdditionalProperties()
  );
}
]